### PR TITLE
timestamp as int96

### PR DESCRIPTION
**Pull request by robert3005**
to maintain backcompat in multi tenant environment

@ash211 @pwoody @darora @jared2501

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala [hunk 1]
```diff
@@ -55,18 +55,21 @@ import org.apache.spark.sql.types._
 private[parquet] class ParquetSchemaConverter(
     assumeBinaryIsString: Boolean = SQLConf.PARQUET_BINARY_AS_STRING.defaultValue.get,
     assumeInt96IsTimestamp: Boolean = SQLConf.PARQUET_INT96_AS_TIMESTAMP.defaultValue.get,
-    writeLegacyParquetFormat: Boolean = SQLConf.PARQUET_WRITE_LEGACY_FORMAT.defaultValue.get) {
+    writeLegacyParquetFormat: Boolean = SQLConf.PARQUET_WRITE_LEGACY_FORMAT.defaultValue.get,
+    writeTimestampAsInt96: Boolean = SQLConf.PARQUET_TIMESTAMP_AS_INT96.defaultValue.get) {
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala]: docs this new param with `@param` above

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala [hunk 2]
```diff
@@ -177,6 +190,18 @@ private[parquet] class ParquetWriteSupport extends WriteSupport[InternalRow] wit
     }
   }
 
+  private def makeTimestampWriter(): ValueWriter = {
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala]: never used?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala [hunk 3]
```diff
@@ -75,6 +84,10 @@ private[parquet] class ParquetWriteSupport extends WriteSupport[InternalRow] wit
       assert(configuration.get(SQLConf.PARQUET_WRITE_LEGACY_FORMAT.key) != null)
       configuration.get(SQLConf.PARQUET_WRITE_LEGACY_FORMAT.key).toBoolean
     }
+    this.writeTimestampAsInt96 = {
+      assert(configuration.get(SQLConf.PARQUET_TIMESTAMP_AS_INT96.key) != null)
+      configuration.get(SQLConf.PARQUET_TIMESTAMP_AS_INT96.key).toBoolean
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala]: need to set this in `ParquetFileFormat#prepareWrite` in the same pattern as `PARQUET_WRITE_LEGACY_FORMAT`?

**Comment by robert3005** [comment]
Thanks @ash211. I was being silly and wrote it too quickly.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -210,6 +210,11 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+  val PARQUET_TIMESTAMP_AS_INT96 = SQLConfigBuilder("spark.sql.parquet.timestampAsInt96")
+    .doc("Write timestamps as int96 to maintain legacy compatibility")
+    .booleanConf
+    .createWithDefault(true)
```
Comment (robert3005) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: 👍

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala [hunk 5]
```diff
@@ -177,6 +188,18 @@ private[parquet] class ParquetWriteSupport extends WriteSupport[InternalRow] wit
     }
   }
 
+  private def makeTimestampWriter(): ValueWriter = {
+    val int96Writer = (row: SpecializedGetters, ordinal: Int) => {
+      val (julianDay, timeOfDayNanos) = DateTimeUtils.toJulianDay(row.getLong(ordinal))
+      val buf = ByteBuffer.wrap(timestampBuffer)
+      buf.order(ByteOrder.LITTLE_ENDIAN).putLong(timeOfDayNanos).putInt(julianDay)
```
Comment (ash211) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala]: I haven't looked at the old code, but hopefully this is exactly what was in the stuff we took out earlier?

**Comment by pwoody** [comment]
@ash211 @robert3005 plz review the test change

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala [hunk 6]
```diff
@@ -757,6 +760,31 @@ class ParquetIOSuite extends QueryTest with ParquetTest with SharedSQLContext {
...
+        withSQLConf(SQLConf.PARQUET_TIMESTAMP_AS_INT96.key -> "false") {
+          readParquetFile(dir.getCanonicalPath) { df =>
+            assert(df.schema.head.dataType == DataTypes.TimestampType)
+            assert(df.collect().map(_.getTimestamp(0)) sameElements
+              timestamps.toArray[Timestamp])
+          }
+        }
+      }
+    }
+  }
```
Comment (ash211) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala]: What do you think of another test for write as int64 and then confirming we can read it back when `PARQUET_TIMESTAMP_AS_INT96` is both true and false?

**Comment by robert3005** [comment]
👍

**Pull request merged**